### PR TITLE
 Delegate Research

### LINKED ISSUE #244: Researcher agent should do research instead of user_proxy
Right now with all of the researcher prompts, the user_proxy plans the research strategy and all the researcher does is call its functions.
This makes it hard to prompt engineer since we would have to give the user_proxy knowledge about token research which should really be reserved for the researcher agent.

This is a bigger issue and likely requires an architecture shift towards nested chats.
Nested chats come with their own set of problems like we discovered with evo.ninja.
The biggest one is: agent A passing context into agent B often leaves out important info or changes details.

### PR DESCRIPTION
**Pull request by nerfZael**
Closes #244

A new `DelegateResearchTokensAgent` is implemented.
It has 1 tool: `research`
With this tool, it can trigger a sub chat between a new user proxy and the old research agent. They will then work on solving the research tasks.

This improves the capability of research tasks, as well as minimizes confusion that the other agents in the group chat can have from viewing research related messages.

However, the execution speed is significantly lower because of a couple of reasons:
- There's an additional delegate call
- There's sometimes an additional summary response from the main user_proxy
- The `DelegateResearchTokensAgent` can be too liberal and sometime use more than one tool call
- There's an additional summarization call
NOTE: in the current implementation I have replaced the summarization LLM call with a message aggregation step to improve performance. While I believe the LLM call can be better (and more succinct) than the aggregation, I have not noticed any negative effects with our current suite of tests.

Further execution speed improvement could be done by running the multiple tool calls (if made) in parallel, but I have not seen an easy way to do that

### CODE REVIEW

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/test_swap_and_send.py::test_send_and_swap_complex 10

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/test_swap_and_send.py::test_send_and_swap_complex 10

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research_and_swap.py,agents/token/research/test_advanced.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research_swaš_and_send.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research_swap_and_send.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research.py

**Comment by nerfZael** [comment]
/workflows/benchmarks agents/token/research/test_research.py

File: autotx/agents/DelegateResearchTokensAgent.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    name: str = "research"
+    description: str = "Research user task"
+
+    def build_tool(self, autotx: AutoTx) -> Callable[[str], str]:
+        def run(
+            tasks: Annotated[str, "User tasks to research"]
+        ) -> str:
+            print(f"Researching user task:", tasks)
+
+            user_proxy_agent = UserProxyAgent(
```
Comment (cbrzn) [inline autotx/agents/DelegateResearchTokensAgent.py:81]: now that we have this new user proxy agent, which takes care of the research, would it make sense to clean up a bit the system message of the OG `user_proxy`?

File: autotx/agents/DelegateResearchTokensAgent.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    More examples:
+    research: {{ "tasks": "What's the 24 hours price change of SHIB?" }}
+    research: {{ "tasks": "Retrieve detailed information on the Ethereum token DAI including current price, market cap, and price change percentages." }}
+    research: {{ "tasks": "Find all tokens available in the DeFi category and filter them based on their performance on the Optimism network." }}
+    research: {{ "tasks": "Compare the price change percentages over the last 24 hours, 7 days, and 30 days for tokens within the Gaming category." }}
+    research: {{ "tasks": "Retrieve a list of all available token categories and then research specific details for the first five tokens in the NFT category sorted by volume in descending order." }}
+    research: {{ "tasks": "Get the Ethereum addresses for all tokens in the Yield Farming category on the Polygon network to prepare for potential trades." }}
+
+    Bad examples:
+    research: {{ "tasks": "Find best crpyto investment strategy" }}
```
Comment (cbrzn) [inline autotx/agents/DelegateResearchTokensAgent.py:41]: typo in crypto

File: autotx/agents/SendTokensAgent.py [hunk 3]
```diff
@@ -46,8 +47,26 @@
...
 
+    Example 3:
+    User: Send 10 USDC to 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045 then buy 5 UNI with ETH and send 40 WBTC to 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
+    Call prepare_transfer_transaction with args:
+    {{
+        "amount": 10,
+        "receiver": "0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045",
+        "token": "USDC"
+    }}
+    NOTE: the second transfer was not prepared because it's waiting for the swap transaction to be prepared first.
```
Comment (cbrzn) [inline autotx/agents/SendTokensAgent.py:58]: the second transfer has nothing to do with the asked swap, right? is that intended?

Comment (nerfZael) [reply autotx/agents/SendTokensAgent.py:58]: Yep, this is just to illustrate to the LLM that the transactions should be created in order as defined by user

File: autotx/agents/DelegateResearchTokensAgent.py (hunk 1 above)
Comment (nerfZael) [reply autotx/agents/DelegateResearchTokensAgent.py:81]: Maybe, but definitely wouldn't want to do it in this PR. Would require a lot of testing for regressions and at best we'd maybe be able to shave off a few sentences

File: autotx/agents/DelegateResearchTokensAgent.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from autotx.autotx_tool import AutoTxTool
+
+name = "research-tokens"
+
+system_message = lambda autotx: dedent(f"""
+    You are an AI assistant that's an expert in Ethereum tokens. Assist the user in their task of researching tokens.
+    ONLY focus on the token research aspect of the user's goal and let other agents handle other tasks.
+    You use the tools available to assist the user in their tasks.  
+    Your job is NOT to buy or sell a token, and there is no need to check the price when doing that.
+    NEVER ask the user questions.
```
[the remark this answers is not in the document]
Comment (nerfZael) [reply autotx/agents/DelegateResearchTokensAgent.py:18]: They essentially have the same task, but different tools. Though I could see the prompt changing depending on their behaviour, which is why I didn't re-use it as a constant

**Review by cbrzn — APPROVED** [review]

**Pull request merged**